design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,367 @@
+
+### Requirement
+
+The solution should support the following capabilities:
+
+1. Deployment, deploy as separate Conjur entity; serve K8s Secrets to many Applications
+2. Rotation, update K8s Secret with Conjur secret values (on trigger or on a timely, configurable basis)
+3. Support removal of deployment
+
+## Solution
+This solution will be divided into milestones.
+
+- Milestone 1: Deployment, deploy as as separate entity
+- Milestone 2: Handle rotation, updating K8s Secrets with Conjur secret value (configurable amount of time or on a trigger)
+
+
+
+### Milestone 1: Deployment
+
+#### Design
+
+The Secrets Provider for K8s will no longer run as an *init* container attached to the application pod, rather run as a separate single Job in a namespace. A Job creates a Pod to perform a task and terminates upon completion. Once the Pod is spun up, it will authenticate to Conjur/DAP in the same way it does today, via authn-k8s. The Pod will then fetch the K8s secret denoted with a label and update them with the Conjur secrets they require.
+
+|      | Deployment                  | Elaboration                                                  | Pros                                 | Cons                                                         |
+| ---- | --------------------------- | ------------------------------------------------------------ | ------------------------------------ | ------------------------------------------------------------ |
+| 1    | Run as Job                  | Run as a K8s Job that deploys a Pod that terminates after task completion | - Doesn't waste customer resources   | - Requires addition of "Job" to application identity granularities. Otherwise a Job will limit us to only using Namespace / Service Account<br /><br />- Deployment will need to be changed (from Job → Pod) for Milestone 2 (support secrets rotation) |
+| 2    | Run as non-terminating  Pod | Run as a Pod that does not terminate automatically after task completion | - Simple<br />- Minimal code changes | - Wastes resources because we aren't updating K8s secrets and up all the time<br /><br />- For **deployment**, **deploymentConfig** and **statefulSet** segregation of duties, this solution is not interchangeable with our other secrets retrieval solutions without policy changes. (e.g. secrets provider as init-container, conjur authn k8s secrets, any of our clients or Conjur REST API) |
+
+**Decision:** Run as a Job
+
+We decided to run as a Job instead of a Pod because for Milestone 1 we are concerned with delivering Conjur values only once. At this time we are not supporting updating K8s Secrets with Conjur secret across multiple runs so deploying a Pod that will run continuously in the environment of a customer is a waste of their resources.
```

Comment:
Why we are not planning ahead? Would we need to change the object type any time we have a new requirement?
We did that mistake the last time and now we need to refactor...

Reply:
the only action it would require by the customer is changing manifest from Job -> Deployment. IMO deploying as a Pod is wasteful especially since the customer can only benefit from the Secrets Provider only once, at the initial spin up
Job:
```
apiVersion: batch/v1
kind: Job
metadata:
  name: secrets-provider
```
Deployment:
```
apiVersion: app/v1
kind: Deployment
metadata:
  name: secrets-provider
```